src/lib/proof-system/zkfunction.ts

```diff
@@ -0,0 +1,185 @@
+import { Snarky, initializeBindings } from '../../snarky.js';
+import { MlFieldArray, MlFieldConstArray } from '../ml/fields.js';
+import { withThreadPool } from '../../snarky.js';
+import { Provable } from '../provable/provable.js';
+import { snarkContext, gatesFromJson } from '../provable/core/provable-context.js';
+import { prettifyStacktrace, prettifyStacktracePromise } from '../util/errors.js';
+import { ProvableTypePure } from '../provable/types/provable-intf.js';
+import { provablePure } from '../provable/types/provable-derivers.js';
+
+// external API
+export { ZkFunction, Keypair, Proof, VerificationKey };
+
+type ZkFunctionConfig<P, W extends any[]> = {
+  name: string;
+  publicInputType: ProvableTypePure<P>;
+  privateInputTypes: { [K in keyof W]: ProvableTypePure<W[K]> };
+  main: (publicInput: P, ...privateInputs: W) => void;
+};
+
+function ZkFunction<P, W extends any[]>(config: ZkFunctionConfig<P, W>) {
+  const publicInputType = provablePure(config.publicInputType);
+  let _keypair: Keypair | undefined;
+
+  return {
+    /**
+     * Generates and stores a proving key and a verification key for this circuit(ZkFunction).
+     *
+     * @returns The generated verification key.
+     *
+     * @example
+     * ```ts
+     * const verificationKey = await zkf.compile();
+     * ```
+     * @warning Must be called before `prove` or `verify`.
+     */
+    async compile() {
+      const main = mainFromCircuitData(config);
+      const publicInputSize = publicInputType.sizeInFields();
+      await initializeBindings();
+      _keypair = await prettifyStacktracePromise(
+        withThreadPool(async () => {
+          let keypair = Snarky.circuit.compile(main, publicInputSize);
+          return new Keypair(keypair);
+        })
+      );
+      return _keypair.verificationKey();
+    },
+
+    /**
+     * Proves a statement using the public input and private inputs of the circuit(ZkFunction).
+     *
+     * @param privateInputs The private inputs to the circuit.
+     * @param publicInput The public input to the circuit.
+     * @returns The generated proof.
+     *
+     * @throws If `compile` has not been called.
+     *
+     * @example
+     * ```ts
+     * await zkf.compile();
+     * const proof = await zkf.prove([privateInput], publicInput);
+     * ```
+     */
+    async prove(privateInputs: W, publicInput: P) {
```

Comment:
nit: `publicInput: P, ...privateInputs: W` would allow us to pass arguments to the prove function like this `await main.prove(hash, preimage);` without having to wrap them in an array